src/ast/rewriter/bit_blaster/bit_blaster_adder.cpp

```diff
@@ -0,0 +1,112 @@
+#include "ast/rewriter/bit_blaster/bit_blaster_adder.h"
+
+bit_blaster_adder::bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, numeral const & value):
+    m_rewriter(rewriter),
+    m_constant(value)
+{
+    reduce();
+    expr_ref_vector empty(m());
+    m_variable.resize(sz, empty);
+}
+
+bit_blaster_adder::bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, expr * const * bits):
+    bit_blaster_adder(rewriter, sz)
+{
+    for (unsigned i = 0; i < sz; i++)
+        add_bit(i, bits[i]);
+}
+
+expr_ref bit_blaster_adder::sum_bits(vector< expr_ref_vector > & columns, expr_ref_vector & out_bits) const {
+    SASSERT(out_bits.empty());
+
+    expr_ref_vector carries(m());
+    expr_ref tmp1(m()), tmp2(m()), tmp3(m());
+
+    for (auto & column : columns) {
+        column.append(carries);
+        carries.reset();
+
+        while (column.size() >= 3) {
+            expr * a = column.back();
+            column.pop_back();
+            expr * b = column.back();
+            column.pop_back();
+            expr * c = column.back();
+            column.pop_back();
+
+            m_rewriter.mk_xor(a, b, tmp1);
```

Comment:
this format is tedious. I finally added ability to tmp1 = m_rewriter.mk_or(a, b);